doc: add helm docs (#32463)
sourcegraph/sourcegraph: doc/admin/install/kubernetes/helm.md

An example of a subchart is shown in the [examples/custom-resources](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples/custom-resources) folder.

More details on how to create and configure a subchart can be found in the [helm documentation](https://helm.sh/docs/chart_template_guide/subcharts_and_globals/).

### Reviewing changes

The `helm template` command can be used to render manifests for review and comparison.

The [Helm Diff plugin](https://github.com/databus23/helm-diff) can also provide a diff against a deployed chart.

## Rendering manifests for kubectl deployment

Manifests rendered using the `helm template` command can be used for direct deployment using `kubectl`.

See [the kubectl documentation](https://kubernetes.io/docs/reference/kubectl/) for more information on using kubectl.

> ## Configuration Options

Reviewer: I'll defer to @jwayman but I don't love having the values documented directly in the docs instead of the repo. It adds a weird extra step to helm PR's since it's a cross-repo change, and makes it so that someone could edit the values in the doc without updating helm.

If we do keep these details in the doc site, probably they should be pulled into a separate page - the list is quite overwhelming.

Author: > It adds a weird extra step to helm PR's since it's a cross-repo change

I don't like this step either.

I am struggling to figure out what part of docs should be kept in docsite and what should belong to the README inside the helm repository.

Maybe Jeff can provide some guidance over here.

# Sourcegraph Helm Chart

## Requirements

* [Helm 3 CLI](https://helm.sh/docs/intro/install/)
* Kubernetes 1.19 or greater

## Quickstart

To use the Helm chart, add Sourcegraph helm repository:
 
```sh
helm repo add sourcegraph https://sourcegraph.github.io/deploy-sourcegraph-helm/
```

Install the Sourcegraph chart using default values:

```sh
helm install sourcegraph sourcegraph/sourcegraph
```

## Configuration guide

Helm customizations can be applied using an override file. Using an override file allows customizations to persist through upgrades without needing to manage merge conflicts.

To customize configuration settings with an override file, create an empty yaml file (e.g. `override.yaml`) and configure overrides.

> > WARNING: __DO NOT__ copy the [default values file](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) as a boilerplate for your override file. You will be risk having outdated values during upgrade.

Reviewer: ```suggestion
> WARNING: __DO NOT__ copy the [default values file](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) as a boilerplate for your override file. You risk having outdated values during upgrades.
```

Example overrides can be found in the [examples](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph/examples) folder. Please take a look at our examples before providing your own configuration and consider using them as boilerplates.

Provide the override file to helm:

```sh
# Installation
helm install --values ./override.yaml sourcegraph sourcegraph/sourcegraph

# Upgrade
helm upgrade --values ./override.yaml sourcegraph sourcegraph/sourcegraph
```

## Configuration options

The Sourcegraph chart is highly customizable to support a wide-range of environment. Please review the default values from [values.yaml](https://github.com/sourcegraph/deploy-sourcegraph-helm/blob/main/charts/sourcegraph/values.yaml) and all [supported options](https://github.com/sourcegraph/deploy-sourcegraph-helm/tree/main/charts/sourcegraph#configuration-options).

## Advanced configuration

The Helm chart is new and still under active development, and we may not cover all your use cases. 

> Please reach out to your account team to discuess your specific need.

Reviewer: ```suggestion
Please reach out to your account team to discuss your specific need.
```

> For advanced users who are looking for a temporary workaround, we __recommend__ applying [Kustomize](https://kustomize.io) on the rendered manifests from our chart. Plesae __do not__ maintain your own fork of our chart, this may impact our ability to support you if you run into issues.

Reviewer: ```suggestion
For advanced users who are looking for a temporary workaround, we __recommend__ applying [Kustomize](https://kustomize.io) on the rendered manifests from our chart. Please __do not__ maintain your own fork of our chart, this may impact our ability to support you if you run into issues.
```
